### PR TITLE
Add documentation regarding non-associativity of >>- fair conjunction.

### PR DESCRIPTION
**Pull request by kquick**

### CODE REVIEW

File: Control/Monad/Logic/Class.hs [hunk 1]
```diff
@@ -120,12 +120,46 @@ class (MonadPlus m) => MonadLogic m where
...
+    --   >>>    \x -> if even x then return x else mzero
+    --
+    --   But the following will /not/ be productive:
+    --
+    --   >>> (return 0 `mplus` return 1) >>-
+    --   >>>    \a -> oddsPlus a >>-
+    --   >>>          \x -> if even x then return x else mzero
+    --
+    --   Since do notation desugaring results in the latter, the
+    --   'RebindableSyntax' language pragma cannot easily be used
```
Comment (Bodigrim) [inline Control/Monad/Logic/Class.hs:159]:
```suggestion
    --   @RebindableSyntax@ language pragma cannot easily be used
```
I don't think pragmas are hyperlinkable.

File: Control/Monad/Logic/Class.hs [hunk 2]
```diff
@@ -120,12 +120,46 @@ class (MonadPlus m) => MonadLogic m where
...
+    --
+    --   >>> (return 0 `mplus` return 1) >>-
+    --   >>>    oddsPlus >>-
+    --   >>>    \x -> if even x then return x else mzero
+    --
+    --   But the following will /not/ be productive:
+    --
+    --   >>> (return 0 `mplus` return 1) >>-
+    --   >>>    \a -> oddsPlus a >>-
+    --   >>>          \x -> if even x then return x else mzero
```
Comment (Bodigrim) [inline Control/Monad/Logic/Class.hs:156]: Please add brackets here and above to make the order of `>>-` explicit for readers.

**Review by Bodigrim — COMMENTED** [review]
Nice!

File: test/Test.hs [hunk 3]
```diff
@@ -260,6 +260,61 @@ main = defaultMain $
...
+        --
+        --   interleave (oddsPlus 0 >>- \x -> if ...)
+        --              (oddsPlus 1 >>- \x -> if ...)
+        --
+        -- which fails to produce any values because interleave still
+        -- requires production of values from both branches to switch
+        -- between those values, but the first (oddsPlus 0 ...) never
+        -- produces any values.
+
+      , expectFail $ testCase "fair conjunction non-productive" $ [2,4,6,8] @=?
```
Comment (Bodigrim) [inline test/Test.hs:300]: `#ifdef MIN_VERSION_tasty....` please

Comment (kquick) [reply test/Test.hs:301]: Whoops!  :-)

**Comment by Bodigrim** [comment]
Awesome, thanks!

**Comment by kquick** [comment]
@Bodigrim , if you have the time it would be nice to get a new Hackage release to get this updated documentation published.

**Comment by Bodigrim** [comment]
@kquick sure. I am trying to switch to `Alternative` vocabulary, but surprisingly it makes timeout tests flaky. 
E. g., https://github.com/Bodigrim/logict/runs/1659764087#step:19:88 - for some reason the timeout was not caught properly by `tasty-expected-failures`.

**Comment by kquick** [comment]
This could be due to a known problem in the interaction between tasty and tasty-expected-failure.  The issue is that timeouts are implemented in tasty itself *outside* of the inner `withTest` framework that tasty-expected-failure has available.  There's a really ugly hack in tasty-expected-failure (I'm allowed to call it that: I wrote it) that sets up a _faster_ timeout so that tasty-expected-failure can catch the timeout before tasty's timer fires.  Unfortunately, this doesn't work very well when the test is highly computational (probably due to where the timeouts are allowed to occur within the GHC RTS) and so the tasty timeout fires and flags the test as a failure.  

You can usually see this for logict tests that are set to 1 second, but the actual test time reported will be significantly longer (on the order of 1.28 or 1.34 seconds or so).  This doesn't precisely match the 1.00s you linked to above, so it's possible this is not the issue, but it's likely.

The tasty author has historically not been open to integrating tasty-expected-failure into the tasty core where this could be addressed in a non-hackish fashion due to (not unreasonable) concerns about complexity and feature creep.  I have a longer term goal of seeing if there might be some modifications to tasty that would allow tasty-expected-failure to be in charge of the timeout process that the tasty author would find reasonable.

tl;dr it's not possible to accurately timeout-fail some computationally intensive test scenarios

**Comment by Bodigrim** [comment]
I've been just reading your comments in `tasty-expected-failures` issue tracker :)

Could you please take a look at the latest master? I switched to `Applicative` / `Alternative` in examples and in the source code where possible. Mostly because `<|>` looks so much better than `` `mplus` ``. If it looks all right, I'll release this week.

**Comment by kquick** [comment]
In general, I do like this better.

* You might want to add some haddock documentation in the main module's header discussing the change from a Monad+MonadPlus implementation to an Applicative+Alternative implementation, and show that
   ```
   (pure 0 <|> pure 1 <|> empty)
   ```
   can be alternatively written as
   ```
   (return 0 `mplus` return 1 `mplus` mzero)
   ```
   This should help bridge the gap between the paper and the current implementation/documentation.

* I also thought about doing the equality test in the examples for associativity of >>-, but decided not to because I thought the numeric values were useful for people to see to understand the way this worked.  Maybe have both?

Other than that, this looks good!

**Comment by Bodigrim** [comment]
Thanks, updated. 

BTW any ideas how to revamp `README.md`?

**Comment by kquick** [comment]
I'm working on something: I'll submit a pull request soon.

I did wonder: with your recent changes to use Applicative/Alternative, does the `MonadLogic` class need the `Monad` constraint, and if not, should it be renamed to `ApplicativeLogic` (with corresponding changes from `msplit` to `asplit`?  Or should it be split into an `ApplicativeLogic` class and a `MonadLogic` class where fair disjunction is in the former and fair conjunction is in the latter?

**Comment by Bodigrim** [comment]
That's an interesting perspective, and indeed it looks appealing to split into `ApplicativeLogic` (I'd say it's rather `AlternativeLogic`) and `MonadLogic`. Are there examples which are the former but not the latter?.. 

But such changes would constitute a major breaking change. I'd like to make improved documentation available in `0.7` series, and ponder about more radical design decisions later.

----

Timeout tests continue to fail from time to time :( This makes it unacceptable for an immediate releases: there are CIs out of our control, e. g., Stackage. I was thinking that instead of setting a global, `tasty` timeout, one can write a special test, which runs two threads: the first one with diverging computation, and the second one with `sleep 10ms`, and succeeds if the second one returns earlier. This would allow to workaround all `tasty` difficulties.

**Comment by kquick** [comment]
Yes, that makes sense to do any `AlternativeLogic` split later because of the breaking API impact.

I just submitted a PR for README enhancements that include an introduction.

That's a bummer about the timeout tests.  That sounds like a good idea: perhaps `race` from the concurrent package would be a good solution until the tasty-expected-failure timeout handling can be improved.  Is that something you were already working on or something you wanted me to take a look at?

**Comment by Bodigrim** [comment]
> I just submitted a PR for README enhancements that include an introduction.

Awesome, I'll take a look tomorrow.

>  Is that something you were already working on or something you wanted me to take a look at?

It would be nice if you could take the lead on this.
